Sort articles

ojcorrigan opened the pull request:
I had already implemented slot of the search functions, most of it is done with calls to the API although I had to do the sort by votes manually with a helper function in utils. The buttons allow the user to sort by author, topic or date and then order ascending or descending.

owen-corrigan committed:
editing another branch

owen-corrigan committed:
created vote up and vote down buttons

owen-corrigan committed:
corrected the decrement that was happening

owen-corrigan committed:
refactored buttons

owen-corrigan committed:
Merging 2 barnches Merge branch 'article-vote'

owen-corrigan committed:
resolve conflict

owen-corrigan committed:
split components into seperate files

owen-corrigan committed:
added sort by votes function

owen-corrigan committed:
have created option to sort articles by votes

ojcorrigan commented:
Turns out I have accidentally done the separate page for each topic in this branch. Uses a link and routes to divert the page to a url with query which can be copy and pasted to get the same results

ojcorrigan commented:
also dealing with error handling on a new branch (some already dealt with just not for topics)

Liam310 commented on src/utils/helper-funcs.jsx:
```diff
@@ -0,0 +1,24 @@
+export const sortFunc = (articlesArr, order) => {
+  function compare(a, b) {
+    const voteA = a.votes;
+    const voteB = b.votes;
+    let comparison = 0;
+    if (!order || order === "ASC") {
+      if (voteA > voteB) {
+        comparison = -1;
+      } else if (voteA < voteB) {
+        comparison = 1;
+      }
+    } else if (order === "DESC") {
+      if (voteA > voteB) {
+        comparison = 1;
+      } else if (voteA < voteB) {
+        comparison = -1;
+      }
+    }
+    return comparison;
+  }
+  return articlesArr.sort(compare);
+};
+
+export default sortFunc;
```
Do you have tests for this helper function? 👀  a great chance to flex those testing skills!

Liam310 commented on src/components/Article-comments.jsx:
```diff
@@ -0,0 +1,81 @@
+    setCommentChange(false);
+    getComments(article_id)
+      .then((data) => {
+        function compare(a, b) {
+          const timeA = a.created_at;
+          const timeB = b.created_at;
+          let comparison = 0;
+          if (timeA > timeB) {
+            comparison = -1;
+          } else if (timeA < timeB) {
+            comparison = 1;
+          }
+          return comparison;
+        }
+        data.comments.sort(compare);
+        setComments(data.comments);
```
Sorting is a very slow process - I would definitely recommend stepping into your BE at another time and implementing the sort functionality there so you can refactor this to not have to worry about it on the FE!

Liam310 commented on src/components/Article-comments.jsx:
```diff
@@ -0,0 +1,81 @@
+import { React, useEffect } from "react";
```
Unnecessary 'React' in the import! (think this might be applicable in some other cases, too...

Liam310 commented on src/components/Articles.jsx:
```diff
@@ -1,23 +1,30 @@
-
+    getArticles(sortBy, order, seeOnly)
+      .then((data) => {
+        if (byVotes) {
+          let sorted = sortFunc(data.articles, order);
+          setArticles(sorted);
```
same as above - using a helper function is a nice workaround in this case, but would be worthy of a refactor at a later point 🙂

Liam310 commented on src/components/Comment-form.jsx:
```diff
@@ -1,11 +1,66 @@
+          } else if (newComment === "Please enter comment") {
+            setIsValid(false);
+          } else if (newComment !== "") {
+            setComments((currComments) => {
+              let comm = {
+                votes: 0,
+                author: username,
+                body: newComment,
+                created_at: Date.now(),
+                comment_id: "TBC",
+              };
+              return [comm, ...currComments];
+            });
+            postComment(newComment, username, article_id);
+            setCommentChange(true);
+            setNewComment("");
+          }
```
This optimistic rendering for comments isn't quite doing what you hope. It is adding the comment into the array, but because you are then calling `setCommentChange` it is triggering the `useEffect` inside the `Article-comments` component, which forces the comments to be fetched again and then re-rendered. This second re-render probably isn't necessary - not the end of the world right now, but something to think about!

Liam310 commented:
hey Owen! Have left some comments on your code 👍

owen-corrigan committed:
addressed comments raised by Liam

ojcorrigan commented:
So I have addressed the issues raised by Liam, I agree a refactor of the back end will be needed. I have written a few tests to make sure my helper function is doing what I wanted. I have also changed the optimistic rendering so that a re-render is not triggered till a refresh. As a result I have disabled the delete comment button until a refresh (just because delete requires a comment_id which I don't have till a re-render)

ojcorrigan commented:
also removed the unwanted reacts( they were a hangover of using a shortcut to create react functions)

owen-corrigan committed:
got rid of some unnessesary state

corkesw commented on src/components/Comment-form.jsx:
```diff
@@ -1,11 +1,69 @@
+      <form
+        className="commentForm"
+        onSubmit={(e) => {
+          e.preventDefault();
```
This logic could be handled with using some of the attributes available to the textarea element - required and placeholder spring to mind

https://www.w3schools.com/tags/tag_textarea.asp

corkesw commented on src/components/Comment-form.jsx:
```diff
@@ -1,11 +1,69 @@
+            postComment(newComment, username, article_id);
+
+            setNewComment("");
+          }
```
We are optimistically rendering the comment but what do you want the UX to be if the comment fails to upload? When do you want to clear the comment in the form? At the moment, if I comment when I'm offline I will lose the form input and have to start again.

corkesw commented on src/components/Comment-form.jsx:
```diff
@@ -1,11 +1,69 @@
+      >
+        <label>Comment: </label>
+        <textarea
+          id="commentField"
```
as above required and placeholder may be useful here

owen-corrigan committed:
addressed stephen's comments

ojcorrigan commented:
Have taken onboard Stephens comments, made it so I'm using required and placeholders for text area, also made it so it's still doesn't re-render the comments on the initial posting, it does wait for a response from the api though that gives me the comment id, so I can now delete the comment before refreshing. Also it leaves the comment in place if it doesn't post. I've got a catch that allows an posting error pop up when an error comes back from the server (tested sending null instead of newComment)

jadelandeg reviewed (approved):
This is looking good. I would also consider adding some sort of indication that things are loading. The articles can take a little bit of time to come up, and can also take a little bit of time to change when changing the sortby etc
